grpc-client-server/prober_server/main.go

```diff
@@ -4,44 +4,116 @@ import (
 	"context"
 	"flag"
 	"fmt"
+	"io"
 	"log"
 	"net"
 	"net/http"
+	"net/http/httptrace"
 	"time"
 
-	pb "github.com/CodeYourFuture/immersive-go-course/grpc-client-server/prober"
+	pb "github.com/Berkeli/immersive-go/grpc-client-server/prober"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/durationpb"
 )
 
 var (
-	port = flag.Int("port", 50051, "The server port")
+	port         = flag.Int("port", 50051, "The server port")
+	LatencyGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "latency_gauge",
+		Help: "The latency of the requests to the endpoint",
+	}, []string{"endpoint", "type"})
+	timeSince = time.Since
 )
 
 // server is used to implement prober.ProberServer.
-type server struct {
+type Server struct {
 	pb.UnimplementedProberServer
 }
 
-func (s *server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
-	// TODO: support a number of repetitions and return average latency
-	start := time.Now()
-	_, _ = http.Get(in.GetEndpoint())	// TODO: add error handling here and check the response code
-	elapsed := time.Since(start)
-	elapsedMsecs := float32(elapsed / time.Millisecond)
+func (s *Server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
+	ttfbTotal := time.Duration(0)
+	ttlbTotal := time.Duration(0)
+	failed := 0
+	for i := 0; i < int(in.GetNumberOfRequests()); i++ {
+		ttfb, ttlb, err := TimedProbe(in.GetEndpoint())
+		if err != nil {
+			log.Printf("could not probe: %v", err)
+			failed++
+			continue
+		}
+		LatencyGauge.WithLabelValues(in.GetEndpoint(), "TTFB").Set(float64(ttfb / time.Millisecond))
+		LatencyGauge.WithLabelValues(in.GetEndpoint(), "TTLB").Set(float64(ttlb / time.Millisecond))
+		ttfbTotal += ttfb
+		ttlbTotal += ttlb
+	}
+	ttfbAverage := time.Duration(float32(ttfbTotal) / float32(in.NumberOfRequests))
+	ttlbAverage := time.Duration(float32(ttlbTotal) / float32(in.NumberOfRequests))
```

Comment:
Do you think we want to average together the TTLB for both successful and unsuccessful responses?

Reply:
Ah good find, we should defenitely not take them into account!